src/credential/presentation/07-evaluate-input-descriptor.ts

```diff
@@ -0,0 +1,235 @@
+        `JSONPath for "${singlePath}" does not match the provided payload.`
+      );
+    }
+    return false;
+  });
+
+  return [matchedPath, matchedValue];
+};
+
+/**
+ * Extracts the claim name from a path that can be in one of the following formats:
+ * 1. $.propertyName
+ * 2. $["propertyName"] or $['propertyName']
+ *
+ * @param path - The path string containing the claim reference.
+ * @returns The extracted claim name if matched; otherwise, throws an exception.
+ */
+const extractClaimName = (path: string): string | undefined => {
+  // Define a regular expression that matches both formats:
+  // 1. $.propertyName
+  // 2. $["propertyName"] or $['propertyName']
+  const regex = /^\$\.(\w+)$|^\$\[(?:'|")(\w+)(?:'|")\]$/;
+
+  const match = path.match(regex);
+  if (match) {
+    // match[1] corresponds to the first capture group (\w+) after $.
+    // match[2] corresponds to the second capture group (\w+) inside [""] or ['']
+    return match[1] || match[2];
+  }
+
+  // If the input doesn't match any of the expected formats, return null
+
+  throw new Error(
+    `Invalid input format: "${path}". Expected formats are "$.propertyName", "$['propertyName']", or '$["propertyName"]'.`
+  );
+};
+
+/**
+ * Categorizes disclosures into required and optional based on claim names and disclosure constraints.
+ *
+ * @param disclosures - An array of DisclosureWithEncoded objects representing the disclosures to categorize.
+ * @param requiredClaimNames - An array of claim names that are marked as required in the input descriptor.
+ * @param optionalClaimNames - An array of claim names that are marked as optional in the input descriptor.
+ * @param isNotLimitDisclosureRequired - A boolean flag indicating whether `limit_disclosure` is not set to "required".
+ *                                      - If `true`, all disclosures are included as optional.
+ *                                      - If `false`, only disclosures matching the required or optional claim names are included.
+ * @returns An object containing two arrays:
+ *          - `requiredDisclosures`: Disclosures that match the required claim names.
+ *          - `optionalDisclosures`: Disclosures that match the optional claim names or all disclosures
+ *                                     if `limit_disclosure` is not required.
+ */
+const categorizeDisclosures = (
+  disclosures: DisclosureWithEncoded[],
+  requiredClaimNames: string[],
+  optionalClaimNames: string[],
+  isNotLimitDisclosure: boolean
+): EvaluatedDisclosures => {
+  const requiredDisclosures: DisclosureWithEncoded[] = [];
+  const optionalDisclosures: DisclosureWithEncoded[] = [];
+
+  disclosures.forEach((disclosure) => {
+    if (requiredClaimNames.includes(disclosure.decoded[1]))
+      requiredDisclosures.push(disclosure);
+    else if (optionalClaimNames.includes(disclosure.decoded[1]))
+      optionalDisclosures.push(disclosure);
+    else if (isNotLimitDisclosure) optionalDisclosures.push(disclosure);
+  });
```

Comment:
Can we use the `filter` function here to make this more readable?

```suggestion
   const requiredDisclosures = disclosures.filter(
    (disclosure) => requiredClaimNames.includes(disclosure.decoded[1])
  );

  const optionalDisclosures = disclosures.filter(
    (disclosure) =>
      optionalClaimNames.includes(disclosure.decoded[1]) ||
      (isNotLimitDisclosure && !requiredClaimNames.includes(disclosure.decoded[1]))
  );
```